torch/csrc/jit/codegen/cuda/lower_shift.h

```diff
@@ -0,0 +1,217 @@
+    return ss.str();
+  }
+
+ private:
+  //! Sizes of the halo regions of two sides. Both values are zero for
+  //! axes with no halo. When an axis has halo at offset zero,
+  //! widths_[0] is non-zero and designates the size of the
+  //! halo. Similarly, non-zero widths_[1] means the axis has halo at
+  //! the other end of the axis.
+  std::array<unsigned int, 2> widths_;
+};
+
+//! Helper class for lowering tensors with halo. Only valid at the
+//! lowering time.
+class HaloInfo {
+ public:
+  //! Scan a fusion and collect all information for lowering
+  void build(Fusion* fusion);
+
+  //! Build mappings of extent information of a TensorDomain
+  void build(TensorDomain* td);
+
+  //! Set initial AxisHaloInfo of a root axis
+  //!
+  //! This is only for root axes. It is an error to query with
+  //! non-root axes.
+  void setRootAxisInfo(IterDomain* id, const AxisHaloInfo& root_axis_info);
+
+  //! Returns the registed AxisHaloInfo of a root axis.
+  //!
+  //! This is only for root axes. It is an error to query with
+  //! non-root axes.
+  const AxisHaloInfo& getRootAxisInfo(IterDomain* id) const;
+  AxisHaloInfo& getRootAxisInfo(IterDomain* id);
+  //! KIR version
+  const AxisHaloInfo& getRootAxisInfo(kir::IterDomain* id) const;
+  AxisHaloInfo& getRootAxisInfo(kir::IterDomain* id);
+
+  //! Query if an axis has a halo width.
+  //!
+  //! See the comment at halo_width_map_.
+  bool hasHaloWidth(IterDomain* id) const;
+
+  //! Return the halo width of an axis.
+  //!
+  //! It's an error if queried for an axis with no halo width
+  //! information.
+  unsigned getHaloWidth(IterDomain* id) const;
+
+  //! Returns an extent if id is extended for halo. Nullptr is
+  //! returned otherwise.
+  Val* getExtent(IterDomain* id) const;
+  kir::Val* getExtent(kir::IterDomain* id) const;
+
+  // True when the extent of id1 is guaranteed to be lesser than or
+  // equal to id2. False when it *may* not.
+  bool extentLessEqual(IterDomain* id1, IterDomain* id2) const;
+  // True when the extent of id1 is guaranteed to be equal to
+  // id2. False when it *may* not.
+  bool extentEqual(IterDomain* id1, IterDomain* id2) const;
+
+  //! Check whether halo predication is needed
+  bool needsShiftPredicate(Expr* expr);
+  bool needsShiftPredicate(kir::Expr* expr);
+
+  std::string toString() const;
+
+ private:
+  //! Propagate root axis information from outputs to inputs of an
+  //! expression
+  void propagateRootAxisInfo(Expr* expr);
+
+  //! Propagate root axis information from consumer to producer
+  void propagateRootAxisInfo(
+      TensorView* producer,
+      TensorView* consumer,
+      Expr* expr);
+
+  //! Validate shift usage
+  void validate(TensorView* td) const;
+
+ private:
+  //! Halo information of root axes
+  std::unordered_map<IterDomain*, AxisHaloInfo> root_axis_map_;
+  //! KIR version
+  std::unordered_map<kir::IterDomain*, AxisHaloInfo> kir_root_axis_map_;
+
+  //! Halo-extended extents. No mapping for normal axes
```

Comment:
It means axes with no halo extension. Not necessarily directly shifted, e.g.:

```
tv0: input tensor of [I, J]
tv1 = tv0 + 1
tv2 = tv1 + 1
tv3 = shift(tv2, {0, 1})
```
Here, `tv1` is not directly shifted, but its second axis would halo extension because of the shift op using `tv2`.

Updated the comment to clarify.